docs/proposals/SIL-Utilities.md

```diff
@@ -0,0 +1,106 @@
+# Utilities for SIL optimizer passes
+This document lists a set of SIL utilities to be used by the Swift implementation of the SIL optimizer.
+
+The purpose of this proposal is to do a better planning of what utilities we need in SIL optimizer passes.
+
+We want to avoid a situation like we have in the C++ SIL optimizer, where a huge amount of (partly) redundant utilities exist. Many of those utilities have overlapping functionality,  are difficult to discover and are poorly designed. We want to do better in the Swift SIL optimizer.
+
+## Basic Data-Structures
+#### `Stack`
+
+An allocation-free, array like data structure. To be used instead of `Swift.Array` wherever random-access is not needed.
+
+**Related C++ utilities:** `llvm::SmallVector`, `Stack`
+**Status:** done  
+
+#### `BasicBlockSet`
+An extremely efficient implementation of a set of basic blocks.
+
+**Related C++ utilities:** `BasicBlockSet`
+**Status:** done  
+
+#### `BasicBlockWorklist`
+To be used for all kind of basic-block work-list algorithms.
+
+**Uses:** `Stack`, `BasicBlockSet`  
+**Related C++ utilities:** `BasicBlockWorklist`
+**Status:** done  
+
+## Building SIL
+
+#### `Builder.insert(after:)`
+Useful to insert instructions after a (potential) terminator instruction.
+
+**Related C++ utilities:** `SILBuilderWithScope::insertAfter()`
+**Status:** done  
+
+## Instruction classifications
+We want to classify certain instructions into common groups so that passes can deal deal with the group instead of individual instruction opcodes. Optimization passes then don't need to be updated if new instructions are added to a group.
+
+In Swift we can easily do this by introducing protocols to which instruction classes can conform to.
+
+#### `ApplySite`
+
+**Related C++ utilities:** `ApplySite`
+**Status:** exists; to-do: complete member variables/functions  
+
+#### `Cast` or `ForwardingCast`
+Details need to be decided.
+
+**Conforming instructions:** `UpcastInst`, `UncheckedRefCastInst`, etc.  
+**Status:** to-do  
+
+#### `AddressProjection`
+Details need to be decided.
+
+**Conforming instructions:** `StructElementAddr`, `TupleElementAddr`  
+**Members:** `var fieldIndex: Int`  
+**Related C++ utilities:** `skipAddrProjections`  
+**Status:** to-do  
+
+## SSA Traversal Utilities
+
+#### Basic Walkers
+This is currently in progress. We'll probably have an up-walkers and  down-walkers for both, addresses and non-address values.
+
+**Uses:** instruction classifications  
+**Related C++ utilities:** `AccessPath`, `RCIdentityAnalysis`, various def-use/use-def walkers in optimization passes.  
+**Status:** in progress
+
+#### `EscapeInfo`
+Escape analysis, which is used e.g. in stack promotion or alias analysis.
+
+**Uses:** basic walkers  
+**Related C++ utilities:** `EscapeAnalysis`, various def-use walkers in optimization passes.  
+**Status:** done, but factoring out the walkers is in progress
+
+#### `AccessPath`
+Details need to be decided.
```

Comment:
I'll provide an `AccessPath` strawman separately.

Reply:
I can added "Appendix: Proposed Swift API" to my latest MemAcessUtils writeup:
https://gist.github.com/atrick/a78780c8476b9503412dabb458dee746#appendix-proposed-swift-api

I can add the information to this document when it lands. I'll somehow distinguish proposed APIs vs. in-use APIs

Reply:
Yes, please open a PR for this. It looks totally reasonable to me, I just have a few minor comments. I think it's best to discuss this in a PR.